README.md
## SynapsePayRest-Ruby

Simple API wrapper for SynapsePay REST V3 API.  This wrapper aims to handle the headers for each API request and appropriate error handling.


## Code Example

Check out the samples.md and our API documentation(http://docs.synapsepay.com/v3.1) for examples.


## Installation

Let `path-to-nowsta-synapse-client.git` be the github address for nowsta's fork of
the synapse rest client.  Add this line to your application's Gemfile:

```ruby
gem 'synapse_pay_rest', git: 'path-to-nowsta-synapse-client.git'
```

And then execute:

    $ bundle


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run
`be rspec` to run the tests. You can also run `bin/console` for an interactive
prompt that will allow you to experiment.


## License

The MIT License (MIT)

Copyright (c) 2014 SynapsePay LLC

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


release_notes
# v0.1.0
* Use synapse unified error format.
* Add option for client to timeout


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "synapse_pay_rest"
require "webmock/rspec"
require "vcr"

RSpec.configure do |c|
  c.disable_monkey_patching!
  c.color = true
  c.order = :random
end

VCR.configure do |c|
  c.cassette_library_dir = "spec/cassettes"
  c.hook_into :webmock
  c.configure_rspec_metadata!
  c.allow_http_connections_when_no_cassette = false
end


spec/synapse_pay_rest_spec.rb
require "spec_helper"

RSpec.describe SynapsePayRest do
  it "has a version number" do
    expect(SynapsePayRest::VERSION).not_to be nil
  end
end


spec/synapse_pay_rest/http_client_spec.rb
require "spec_helper"

RSpec.describe SynapsePayRest::HTTPClient do
  let(:client_id) { "e3f19e4bd4022c86e7f2" }
  let(:client_secret) { "11c94ba6bad74d24a0158bc707f0fc19a86dc08f" }
  let(:ip_address) { "107.170.246.225" }
  let(:fingerprint) { "e716990e50b67a1177736960b6357524b22090ccab093d068b3d7a18dbde3f4c" }

  let(:url_base) { "https://sandbox.synapsepay.com/api/3" }
  let(:path) { "/users" }
  let(:base_params) do
    {
      "fingerprint" => fingerprint,
      "client_id" => client_id,
      "client_secret" => client_secret,
      "ip_address" => ip_address,
    }
  end
  let(:params) { base_params }
  let(:client) { SynapsePayRest::HTTPClient.new(params, url_base) }

  describe "#get" do
    let(:result) { client.get(path) }

    # Basic error handling is tested through get requests.
    # All http methods use the same error handling mechanism so when possible
    # it is best to try to keep most of the error handling code in this
    # context.
    context "when there are errors" do
      context "when the synapse response is a well formed error", :vcr do
        let(:path) { "/users/xyz" }

        it { expect(result["http_code"]).to eq("404") }
        it { expect(result["error_code"]).to eq("404") }
        it { expect(result["error"]["en"]).to be_a(String) }
        it { expect(result["success"]).to eq(false) }
      end

      context "when the synapse response is not a well formed error" do
        before(:each) do
          # error result of a 5XX error producing html pulled from bugsnag
          # https://bugsnag.com/nowsta-1/pay/errors/576c0020e8a4560f523354b3
          body = '<html><head><title>504 Gateway Time-out</title></head>' \
            '<body><h1>504 Gateway Time-out</h1></body></html>'
          stub_request(:any, /#{url_base}/).to_return(body: body, status: 504)
        end

        it { expect(result["http_code"]).to eq("504") }
        it { expect(result["error_code"]).to eq("500") }
        it { expect(result["error"]["en"]).to be_a(String) }
        it { expect(result["success"]).to eq(false) }
      end

      context "when client timeout is reached" do
        before(:each) do
          msg = "Timed out connecting to server."
          allow(RestClient::Request).to receive(:execute).with(
            hash_including(timeout: timeout)
          ).and_raise(RestClient::Exceptions::ReadTimeout.new(msg))
        end
        let(:timeout) { 0.4 }
        let(:params) { base_params.merge("timeout" => timeout) }

        it { expect(result["http_code"]).to eq("504") }
        it { expect(result["error_code"]).to eq("500") }
        it { expect(result["error"]["en"]).to be_a(String) }
        it { expect(result["success"]).to eq(false) }
      end
    end

    context "happy path", :vcr do
      it { expect(result["error_code"]).to eq("0") }
      it { expect(result["success"]).to eq(true) }
    end
  end

  # since all the error testing id done by the get request, we just need
  # to check that the rest client is getting called with the correct inputs.
  def self.it_makes_request_with_body(method:)
    it "makes request with with body" do
      response = { "expected" => "response" }
      allow(RestClient::Request).to receive(:execute).and_return(response.to_json)

      payload = { "expected" => "payload" }
      expect(client.send(method, path, payload)).to eq(response)
      expect(RestClient::Request).to have_received(:execute).with(
        hash_including(method: method, url: url_base + path, payload: payload.to_json)
      )
    end
  end

  describe "#post" do
    it_makes_request_with_body method: :post
  end

  describe "#patch" do
    it_makes_request_with_body method: :patch
  end

  describe "#delete" do
    it "makes delete request" do
      response = { "expected" => "response" }
      allow(RestClient::Request).to receive(:execute).and_return(response.to_json)

      expect(client.delete(path)).to eq(response)
      expect(RestClient::Request).to have_received(:execute).with(
        hash_including(method: :delete, url: url_base + path)
      )
    end
  end
end


lib/synapse_pay_rest.rb
require "synapse_pay_rest/version"

# Basic wrapper around the the requests library.
require_relative "synapse_pay_rest/http_client"
# Assign all the api classes
require_relative "synapse_pay_rest/api/Users"
require_relative "synapse_pay_rest/api/Nodes"
require_relative "synapse_pay_rest/api/Trans"

module SynapsePayRest
  class Client

    attr_accessor :client
    attr_accessor :users
    attr_accessor :nodes
    attr_accessor :trans

    def initialize(options: raise("options is required"), user_id: nil)
      base_url = 'https://synapsepay.com/api/3'
      if options.has_key?('development_mode')
        if options['development_mode']
          base_url = 'https://sandbox.synapsepay.com/api/3'
        end
      end

      @client = HTTPClient.new options, base_url, user_id: user_id
      @users = Users.new @client
      @nodes = Nodes.new @client
      @trans = Trans.new @client
    end
  end
end


lib/synapse_pay_rest/api/Users.rb
require 'mime-types'
require 'base64'
require 'open-uri'

module SynapsePayRest
  class Users

    attr_accessor :client

    def initialize(client)
      @client = client
    end

    def create_user_path(user_id: nil)
      path = '/users'
      if user_id
        path += '/%s' %user_id
      end
      return path
    end

    def refresh(payload: raise("payload is required"))
      path = '/oauth/%s' % @client.user_id
      response = @client.post(path, payload)
      if response.has_key?('oauth_key')
        @client.update_headers(oauth_key: response['oauth_key'])
      end
      return response
    end

    def get(user_id: nil, query: nil, page: nil, per_page: nil)
      if user_id
        path = create_user_path(user_id: user_id)
        response = @client.get(path)
        if response.has_key?('_id')
          @client.update_headers(user_id: response['_id'])
        end
      elsif query
        path += '?query=%s' %query
        if page
          path += '&page=%s' %page
        end
        if per_page
          path += '&per_page=%s' %per_page
        end
      elsif page
        path += '?page=%s' %page
        if per_page
          path += '&per_page=%s' %per_page
        end
      elsif per_page
        path += '?page=%s' %per_page
      else
        path = create_user_path()
      end
      response = @client.get(path)
      return response
    end

    def update(payload: raise("payload is required"))
      path = create_user_path(user_id: @client.user_id)
      response = @client.patch(path, payload)
      return response
    end

    def create(payload: raise("payload is required"))
      path = create_user_path()
      response = @client.post(path, payload)
      if response.has_key?('_id')
        @client.update_headers(user_id: response['_id'])
      end
      return response
    end

    def add_doc(payload: raise("payload is required"))
      path = create_user_path(user_id: @client.user_id)
      response = @client.patch(path, payload)
      return response
    end

    def answer_kba(payload: raise("payload is required"))
      path = create_user_path(user_id: @client.user_id)
      response = @client.patch(path, payload)
      return response
    end

    def attach_file(file_path: raise("file_path is required"))
      path = create_user_path(user_id: @client.user_id)
      file_contents = open(file_path) { |f| f.read }
      file_type = MIME::Types.type_for(file_path).first.content_type
      mime_padding = 'data:' + file_type + ';base64,'
      encoded = Base64.encode64(file_contents)
      base64_attachment = mime_padding + encoded
      payload = {
        'doc' => {
          'attachment' => base64_attachment
        }
      }
      response = @client.patch(path, payload)
      return response
    end
  end
end


lib/synapse_pay_rest/http_client.rb
require 'rest-client'
require 'json'

module SynapsePayRest
  class HTTPClient

    attr_accessor :base_url
    attr_accessor :options
    attr_accessor :headers
    attr_accessor :user_id
    attr_accessor :timeout

    def initialize(options, base_url, user_id: nil)
      @options = options
      user = '|%s' % options['fingerprint']
      if options.has_key?('oauth_key')
        user = '%s|%s' % [options['oauth_key'], options['fingerprint']]
      end
      gateway = '%s|%s' % [options['client_id'], options['client_secret']]
      @headers = {
        :content_type => :json,
        :accept => :json,
        'X-SP-GATEWAY' => gateway,
        'X-SP-USER' => user,
        'X-SP-USER-IP' => options['ip_address']
      }
      @base_url = base_url
      # RestClient.log = 'stdout'
      @user_id = user_id
      @timeout = options['timeout']
    end

    def update_headers(user_id: nil, oauth_key: nil, fingerprint: nil, client_id: nil, client_secret: nil, ip_address: nil)
      if user_id
        @user_id = user_id
      end
      if oauth_key and !fingerprint
        @headers['X-SP-USER'] = '%s|%s' % [oauth_key, @options['fingerprint']]
      elsif oauth_key and fingerprint
        @headers['X-SP-USER'] = '%s|%s' % [oauth_key, fingerprint]
      end

      if client_id and !client_secret
        @headers['X-SP-GATEWAY'] = '%s|%s' % [client_id, @options['client_secret']]
      elsif client_id and client_secret
        @headers['X-SP-GATEWAY'] = '%s|%s' % [client_id, client_secret]
      elsif !client_id and client_secret
        @headers['X-SP-GATEWAY'] = '%s|%s' % [@options['client_id'], client_secret]
      end

      if ip_address
        @headers['X-SP-USER-IP'] = ip_address
      end
    end

    def post(path, payload)
      url = base_url + path
      response = with_error_handling do
        make_request(method: :post, url: url, payload: payload.to_json)
      end
      JSON.parse(response)
    end

    def patch(path, payload)
      url = base_url + path
      response = with_error_handling do
        make_request(method: :patch, url: url, payload: payload.to_json)
      end
      JSON.parse(response)
    end

    def get(path)
      url = base_url + path
      response = with_error_handling { make_request(method: :get, url: url) }
      JSON.parse(response)
    end

    def delete(path)
      url = base_url + path
      response = with_error_handling { make_request(method: :delete, url: url) }
      JSON.parse(response)
    end

    private

    def make_request(options)
      base_options = { headers: @headers }
      base_options[:timeout] = timeout if timeout
      RestClient::Request.execute(base_options.merge(options))
    end

    def with_error_handling
      yield
    rescue RestClient::RequestTimeout => e
      format_error(504, messages[:timeout])
    rescue RestClient::ExceptionWithResponse => e
      code = e.response.code
      well_formed?(code) ? e.response : format_error(code, messages[:error])
    end

    def well_formed?(code)
      # synapse docs lists 500 as a well formed, however, on 1 July 2016
      # we received:
      #   <h1>Server Error (500)</h1>
      # when testing throttling.
      [400, 401, 402, 404, 409, 429].include?(code)
    end

    def messages
      map = {
        timeout: "Our payments service was unresponsive.",
        error: "Error occurred with our payments service.",
      }
      map.default = "An unhanded error occurred with our payments service."
      map
    end

    def format_error(code, message)
      {
        "error_code" => "500",
        "http_code" => code.to_s,
        "error" => { "en" => message },
        "success" => false,
      }.to_json
    end
  end
end
